src/compiler/checker.ts

```diff
@@ -34560,12 +34564,17 @@ namespace ts {
                 return;
             }
 
+            const testedType = location === condExpr ? type : checkExpression(location);
+            if (typeToString(testedType).includes('undefined')) { // maybeTypeOfKind(testedType, TypeFlags.Undefined)

```

Comment:
I don't really understand why, but `maybeTypeOfKind(testedType, TypeFlags.Undefined)` always returns `false`, even for types like `() => boolean | undefined`. This is because for some reason that I am unaware of the type ``() => boolean | undefined`` is lacking the `Union` flag. So what I put into the `if` here is just a placeholder.